BUG: optimize: broken array types support in `least_squares`

After gh-22728, running the `optimize` tests with `SCIPY_ARRAY_API=1` set yields 56 failures containing:

```
scipy/_lib/_array_api.py:160: in array_namespace
    api_arrays = list(_compliance_scipy(arrays))
        arrays     = (<Compressed Sparse Row sparse array of dtype 'float64'
        with 6 stored elements and shape (2, 6)>,)
scipy/_lib/_array_api.py:87: in _compliance_scipy
    raise ValueError(msg)
E   ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `scipy.sparse.linalg` functions would work instead.
```

Full list of failures and example traceback:

<details>

```
________________ TestTrustRegionConstr.test_list_of_problems[hess4-False-prob15] _________________
scipy/optimize/tests/test_minimize_constrained.py:489: in test_list_of_problems
    result = minimize(prob.fun, prob.x0,
        grad       = False
        hess       = <scipy.optimize._hessian_update_strategy.BFGS object at 0x7d8358c80b30>
        prob       = <scipy.optimize.tests.test_minimize_constrained.Elec object at 0x7d8358f024e0>
        self       = <scipy.optimize.tests.test_minimize_constrained.TestTrustRegionConstr object at 0x7d8358b60f80>
        sensitive  = False
        sup        = <numpy.testing._private.utils.suppress_warnings object at 0x7d8358b3db50>
scipy/optimize/_minimize.py:763: in minimize
    res = _minimize_trustregion_constr(fun, x0, args, jac, hess, hessp,
        args       = ()
        bounds     = None
        callback   = None
        constraints = [<scipy.optimize._constraints.NonlinearConstraint object at 0x7d8358b3c290>]
        fun        = <bound method Elec.fun of <scipy.optimize.tests.test_minimize_constrained.Elec object at 0x7d8358f024e0>>
        hess       = <scipy.optimize._hessian_update_strategy.BFGS object at 0x7d8358c80b30>
        hessp      = None
        jac        = '2-point'
        meth       = 'trust-constr'
        method     = 'trust-constr'
        options    = {}
        remove_vars = False
        tol        = None
        x0         = array([-0.76141812, -0.20841074, -0.24113917, -0.93761493,  0.60174275,
        0.27828619])
scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py:544: in _minimize_trustregion_constr
    _, result = tr_interior_point(
        J_eq0      = <Compressed Sparse Row sparse array of dtype 'float64'
        with 0 stored elements and shape (0, 6)>
        J_ineq0    = <Compressed Sparse Row sparse array of dtype 'float64'
        with 6 stored elements and shape (2, 6)>
        args       = ()
        barrier_tol = 1e-08
        bounds     = None
        c_eq0      = array([], dtype=float64)
        c_ineq0    = array([0., 0.])
        callback   = None
        canonical  = <scipy.optimize._trustregion_constr.canonical_constraint.CanonicalConstraint object at 0x7d8364b303e0>
        canonical_all = [<scipy.optimize._trustregion_constr.canonical_constraint.CanonicalConstraint object at 0x7d8364b303e0>]
        constraints = [<scipy.optimize._constraints.NonlinearConstraint object at 0x7d8358b3c290>]
        disp       = False
        factorization_method = None
        finite_diff_bounds = (-inf, inf)
        finite_diff_rel_step = None
        fun        = <bound method Elec.fun of <scipy.optimize.tests.test_minimize_constrained.Elec object at 0x7d8358f024e0>>
        grad       = '2-point'
        gtol       = 1e-08
        hess       = <scipy.optimize._hessian_update_strategy.BFGS object at 0x7d8358c80b30>
        hessp      = None
        initial_barrier_parameter = 0.1
        initial_barrier_tolerance = 0.1
        initial_constr_penalty = 1.0
        initial_tr_radius = 1.0
        lagrangian_hess = <scipy.optimize._trustregion_constr.minimize_trustregion_constr.LagrangianHessian object at 0x7d8364b30cb0>
        maxiter    = 1000
        method     = 'tr_interior_point'
        n_sparse   = 1
        n_vars     = 6
        objective  = <scipy.optimize._differentiable_functions.ScalarFunction object at 0x7d8358b3eb10>
        prepared_constraints = [<scipy.optimize._constraints.PreparedConstraint object at 0x7d8364b30fe0>]
        sparse_jacobian = True
        start_time = 1743524360.3151567
        state      =             status: None
               fun: 1.0567260307031066
                 x: [-7.614e-01 -2.084e-01 -2.411e-01 ...4
         tr_radius: 1.0
    constr_penalty: 1.0
 barrier_parameter: 0.1
 barrier_tolerance: 0.1
             niter: 1
        stop_criteria = <function _minimize_trustregion_constr.<locals>.stop_criteria at 0x7d83622ab740>
        verbose    = 0
        workers    = None
        x0         = array([-0.76141812, -0.20841074, -0.24113917, -0.93761493,  0.60174275,
        0.27828619])
        xtol       = 1e-08
scipy/optimize/_trustregion_constr/tr_interior_point.py:336: in tr_interior_point
    z, state = equality_constrained_sqp(
        BARRIER_DECAY_RATIO = 0.2
        BOUNDARY_PARAMETER = 0.995
        TRUST_ENLARGEMENT = 5
        barrier_parameter = 0.1
        constr     = <function CanonicalConstraint._less_to_canonical.<locals>.fun at 0x7d83622ab560>
        constr0_subprob = array([1., 1.])
        constr_eq0 = array([], dtype=float64)
        constr_ineq0 = array([0., 0.])
        enforce_feasibility = array([False, False])
        factorization_method = None
        finite_diff_bounds = (-inf, inf)
        fun        = <bound method ScalarFunction.fun of <scipy.optimize._differentiable_functions.ScalarFunction object at 0x7d8358b3eb10>>
        fun0       = np.float64(1.0567260307031066)
        fun0_subprob = np.float64(1.0567260307031066)
        grad       = <bound method ScalarFunction.grad of <scipy.optimize._differentiable_functions.ScalarFunction object at 0x7d8358b3eb10>>
        grad0      = array([ 0.65255652, -0.65255655, -0.82185127,  0.82185124, -0.38168332,
        0.38168332])
        grad0_subprob = array([ 0.65255652, -0.65255655, -0.82185127,  0.82185124, -0.38168332,
        0.38168332, -0.1       , -0.1       ])
        initial_barrier_parameter = 0.1
        initial_penalty = 1.0
        initial_tolerance = 0.1
        initial_trust_radius = 1.0
        jac        = <function CanonicalConstraint._less_to_canonical.<locals>.jac at 0x7d83622ab240>
        jac0_subprob = <Compressed Sparse Row sparse array of dtype 'float64'
        with 8 stored elements and shape (2, 8)>
        jac_eq0    = <Compressed Sparse Row sparse array of dtype 'float64'
        with 0 stored elements and shape (0, 6)>
        jac_ineq0  = <Compressed Sparse Row sparse array of dtype 'float64'
        with 6 stored elements and shape (2, 6)>
        lagr_hess  = <scipy.optimize._trustregion_constr.minimize_trustregion_constr.LagrangianHessian object at 0x7d8364b30cb0>
        n_eq       = 0
        n_ineq     = np.int64(2)
        n_vars     = 6
        s0         = array([1., 1.])
        state      =             status: None
               fun: 1.0567260307031066
                 x: [-7.614e-01 -2.084e-01 -2.411e-01 ...4
         tr_radius: 1.0
    constr_penalty: 1.0
 barrier_parameter: 0.1
 barrier_tolerance: 0.1
             niter: 1
        stop_criteria = <function _minimize_trustregion_constr.<locals>.stop_criteria at 0x7d83622ab740>
        subprob    = <scipy.optimize._trustregion_constr.tr_interior_point.BarrierSubproblem object at 0x7d8364b33920>
        tolerance  = 0.1
        trust_lb   = array([  -inf,   -inf,   -inf,   -inf,   -inf,   -inf, -0.995, -0.995])
        trust_radius = 1.0
        trust_ub   = array([inf, inf, inf, inf, inf, inf, inf, inf])
        x0         = array([-0.76141812, -0.20841074, -0.24113917, -0.93761493,  0.60174275,
        0.27828619])
        xtol       = 1e-08
        z          = array([-0.76141812, -0.20841074, -0.24113917, -0.93761493,  0.60174275,
        0.27828619,  1.        ,  1.        ])
scipy/optimize/_trustregion_constr/equality_constrained_sqp.py:214: in equality_constrained_sqp
    c, A = grad_and_jac(x)
        A          = <Compressed Sparse Row sparse array of dtype 'float64'
        with 8 stored elements and shape (2, 8)>
        BOX_FACTOR = 0.5
        H          = <8x8 _CustomLinearOperator with dtype=float64>
        INTERMEDIARY_REDUCTION_RATIO = 0.3
        LARGE_REDUCTION_RATIO = 0.9
        LS         = <2x8 _CustomLinearOperator with dtype=float64>
        MAX_TRUST_REDUCTION = 0.5
        MIN_TRUST_REDUCTION = 0.1
        PENALTY_FACTOR = 0.3
        S          = <8x8 _CustomLinearOperator with dtype=float64>
        SOC_THRESHOLD = 0.1
        SUFFICIENT_REDUCTION_RATIO = 1e-08
        TRUST_ENLARGEMENT_FACTOR_L = 7.0
        TRUST_ENLARGEMENT_FACTOR_S = 2.0
        TR_FACTOR  = 0.8
        Y          = <8x2 _CustomLinearOperator with dtype=float64>
        Z          = <8x8 _CustomLinearOperator with dtype=float64>
        actual_reduction = np.float64(1.2454747563222701)
        b          = array([0.3329613, 0.3329613])
        b_next     = array([0.3329613, 0.3329613])
        b_t        = array([0., 0.])
        c          = array([ 0.65255652, -0.65255655, -0.82185127,  0.82185124, -0.38168332,
        0.38168332, -0.1       , -0.1       ])
        c_t        = array([ 1.26169102, -0.48582796, -0.62893993,  1.57194318, -0.86307752,
        0.15905437, -0.14626904, -0.14626904])
        cg_info    = {'hits_boundary': True, 'niter': 1, 'stop_cond': 2}
        constr0    = array([1., 1.])
        constr_violation = np.float64(1.0)
        d          = array([ 0.09905683,  0.44578513,  0.54944777,  0.11276669, -0.14579505,
       -0.34859803, -0.40870368, -0.40870368])
        dn         = array([ 0.30456725,  0.0833643 ,  0.09645567,  0.37504597, -0.2406971 ,
       -0.11131448, -0.2       , -0.2       ])
        dt         = array([-0.20551042,  0.36242084,  0.4529921 , -0.26227929,  0.09490205,
       -0.23728355, -0.20870368, -0.20870368])
        f          = np.float64(0.754586451219859)
        f_next     = np.float64(0.754586451219859)
        factorization_method = None
        fun0       = np.float64(1.0567260307031066)
        fun_and_constr = <bound method BarrierSubproblem.function_and_constraints of <scipy.optimize._trustregion_constr.tr_interior_point.BarrierSubproblem object at 0x7d8364b33920>>
        grad0      = array([ 0.65255652, -0.65255655, -0.82185127,  0.82185124, -0.38168332,
        0.38168332, -0.1       , -0.1       ])
        grad_and_jac = <bound method BarrierSubproblem.gradient_and_jacobian of <scipy.optimize._trustregion_constr.tr_interior_point.BarrierSubproblem object at 0x7d8364b33920>>
        initial_penalty = 1.0
        initial_trust_radius = 1.0
        jac0       = <Compressed Sparse Row sparse array of dtype 'float64'
        with 8 stored elements and shape (2, 8)>
        lagr_hess  = <bound method BarrierSubproblem.lagrangian_hessian of <scipy.optimize._trustregion_constr.tr_interior_point.BarrierSubproblem object at 0x7d8364b33920>>
        last_iteration_failed = False
        lb_t       = array([  -inf,   -inf,   -inf,   -inf,   -inf,   -inf, -0.795, -0.795])
        linearized_constr = array([0., 0.])
        merit_function = np.float64(2.470939593076202)
        merit_function_next = np.float64(1.2254648367539318)
        n          = 8
        new_penalty = np.float64(0.12501010240790525)
        optimality = np.float64(0.933424048749457)
        penalty    = 1.0
        predicted_reduction = np.float64(1.2904598747918588)
        previous_penalty = 1.0
        quadratic_model = np.float64(0.1237536875812364)
        reduction_ratio = np.float64(0.9651402423675944)
        scaling    = <bound method BarrierSubproblem.scaling of <scipy.optimize._trustregion_constr.tr_interior_point.BarrierSubproblem object at 0x7d8364b33920>>
        state      =             status: None
               fun: 1.0567260307031066
                 x: [-7.614e-01 -2.084e-01 -2.411e-01 ...4
         tr_radius: 1.0
    constr_penalty: 1.0
 barrier_parameter: 0.1
 barrier_tolerance: 0.1
             niter: 1
        stop_criteria = <bound method BarrierSubproblem.stop_criteria of <scipy.optimize._trustregion_constr.tr_interior_point.BarrierSubproblem object at 0x7d8364b33920>>
        trust_lb   = array([  -inf,   -inf,   -inf,   -inf,   -inf,   -inf, -0.995, -0.995])
        trust_radius = np.float64(7.0)
        trust_radius_t = np.float64(0.7745966692414833)
        trust_ub   = array([inf, inf, inf, inf, inf, inf, inf, inf])
        ub_t       = array([inf, inf, inf, inf, inf, inf, inf, inf])
        v          = array([0.2313452, 0.2313452])
        vpred      = np.float64(1.4142135623730951)
        x          = array([-0.6623613 ,  0.23737439,  0.30830859, -0.82484824,  0.4559477 ,
       -0.07031184,  0.59129632,  0.59129632])
        x0         = array([-0.76141812, -0.20841074, -0.24113917, -0.93761493,  0.60174275,
        0.27828619,  1.        ,  1.        ])
        x_next     = array([-0.6623613 ,  0.23737439,  0.30830859, -0.82484824,  0.4559477 ,
       -0.07031184,  0.59129632,  0.59129632])
scipy/optimize/_trustregion_constr/tr_interior_point.py:147: in gradient_and_jacobian
    J_eq, J_ineq = self.jac(x)
        g          = array([ 0.24651884, -0.24651884, -0.31047396,  0.31047396, -0.14419   ,
        0.14419   ])
        s          = array([0.59129632, 0.59129632])
        self       = <scipy.optimize._trustregion_constr.tr_interior_point.BarrierSubproblem object at 0x7d8364b33920>
        x          = array([-0.6623613 ,  0.23737439,  0.30830859, -0.82484824,  0.4559477 ,
       -0.07031184])
        z          = array([-0.6623613 ,  0.23737439,  0.30830859, -0.82484824,  0.4559477 ,
       -0.07031184,  0.59129632,  0.59129632])
scipy/optimize/_trustregion_constr/canonical_constraint.py:201: in jac
    return empty_jac, cfun.jac(x)
        cfun       = <scipy.optimize._differentiable_functions.VectorFunction object at 0x7d8364b305c0>
        empty_jac  = <Compressed Sparse Row sparse array of dtype 'float64'
        with 0 stored elements and shape (0, 6)>
        x          = array([-0.6623613 ,  0.23737439,  0.30830859, -0.82484824,  0.4559477 ,
       -0.07031184])
scipy/optimize/_differentiable_functions.py:743: in jac
    xp = array_namespace(self.J)
        self       = <scipy.optimize._differentiable_functions.VectorFunction object at 0x7d8364b305c0>
        x          = array([-0.6623613 ,  0.23737439,  0.30830859, -0.82484824,  0.4559477 ,
       -0.07031184])
scipy/_lib/_array_api.py:160: in array_namespace
    api_arrays = list(_compliance_scipy(arrays))
        arrays     = (<Compressed Sparse Row sparse array of dtype 'float64'
        with 6 stored elements and shape (2, 6)>,)
scipy/_lib/_array_api.py:87: in _compliance_scipy
    raise ValueError(msg)
E   ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `scipy.sparse.linalg` functions would work instead.
        array      = <Compressed Sparse Row sparse array of dtype 'float64'
        with 6 stored elements and shape (2, 6)>
        arrays     = (<Compressed Sparse Row sparse array of dtype 'float64'
        with 6 stored elements and shape (2, 6)>,)
        msg        = 'Sparse arrays/matrices are not supported by this function. Perhaps one of the `scipy.sparse.linalg` functions would work instead.'
==================================== short test summary info =====================================
FAILED scipy/optimize/_trustregion_constr/tests/test_canonical_constraint.py::test_nonlinear_constraint - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/_trustregion_constr/tests/test_canonical_constraint.py::test_concatenation - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestDogbox::test_equivalence - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestDogbox::test_tr_options - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestDogbox::test_wrong_parameters - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestDogbox::test_solver_selection - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestDogbox::test_numerical_jac - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestDogbox::test_with_bounds - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestDogbox::test_x_scale_jac_scale - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestTRF::test_equivalence - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestTRF::test_tr_options - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestTRF::test_wrong_parameters - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestTRF::test_solver_selection - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestTRF::test_numerical_jac - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestTRF::test_with_bounds - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestTRF::test_x_scale_jac_scale - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_least_squares.py::TestTRF::test_lsmr_regularization - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-prob.grad-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-prob.grad-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-prob.grad-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-3-point-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-3-point-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-3-point-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-False-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-False-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[prob.hess-False-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[3-point-prob.grad-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[3-point-prob.grad-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[3-point-prob.grad-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-prob.grad-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-prob.grad-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-prob.grad-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-3-point-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-3-point-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-3-point-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-False-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-False-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess2-False-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-prob.grad-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-prob.grad-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-prob.grad-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-3-point-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-3-point-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-3-point-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-False-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-False-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess3-False-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-prob.grad-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-prob.grad-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-prob.grad-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-3-point-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-3-point-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-3-point-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-False-prob13] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-False-prob14] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
FAILED scipy/optimize/tests/test_minimize_constrained.py::TestTrustRegionConstr::test_list_of_problems[hess4-False-prob15] - ValueError: Sparse arrays/matrices are not supported by this function. Perhaps one of the `sc...
= 56 failed, 3362 passed, 400 skipped, 7 deselected, 13 xfailed, 3 xpassed in 106.85s (0:01:46) ==
```

</details>

This also demonstrates that in the process of trying to make the array API tests run faster by selecting tests with `-m array_api_backends` (which selects tests using the `xp` fixture), we lost test coverage for the regular test suite with `SCIPY_ARRAY_API=1`. We should fix that by setting that environment variable in some other test job; we want the whole test suite with regular `numpy` to pass, since it has different (future) input validation behavior. We should pick a job using `-m full`, e.g. in `linux.yml`.

--- a/.github/workflows/linux.yml
+++ b/.github/workflows/linux.yml
@@ -285,7 +285,7 @@ jobs:
   #################################################################################
   prerelease_deps_coverage_64bit_blas:
     # TODO: re-enable ILP64 build.
-    name: Prerelease deps & coverage report, full, py3.11/npMin & py3.11/npPre, dev.py
+    name: Prerelease deps & coverage report, full, py3.11/npMin & py3.11/npPre, dev.py, SCIPY_ARRAY_API=1
     needs: get_commit_message
     if: >
       needs.get_commit_message.outputs.message == 1
@@ -362,6 +362,7 @@ jobs:
     - name: Test SciPy
       run: |
         export OPENBLAS_NUM_THREADS=1
+        export SCIPY_ARRAY_API=1
         python dev.py --no-build test --coverage -j2 --mode full -- --cov --cov-report term-missing
 
   #################################################################################

--- a/scipy/optimize/_differentiable_functions.py
+++ b/scipy/optimize/_differentiable_functions.py
@@ -622,6 +622,8 @@ class VectorFunction:
         self.sparse_jacobian = False
         if (sparse_jacobian or
                 sparse_jacobian is None and sps.issparse(self.J)):
+            # something truthy was specified for sparse_jacobian,
+            # or it turns out that the Jacobian was sparse.
             self.J = sps.csr_array(self.J)
             self.sparse_jacobian = True
         elif sps.issparse(self.J):
@@ -740,8 +742,7 @@ class VectorFunction:
         if hasattr(self.J, "astype"):
             # returns a copy so that downstream can't overwrite the
             # internal attribute. But one can't copy a LinearOperator
-            xp = array_namespace(self.J)
-            return xp.astype(self.J, self.J.dtype)
+            return self.J.astype(self.J.dtype)
         return self.J
 
     def hess(self, x, v):
@@ -752,8 +753,7 @@ class VectorFunction:
         if hasattr(self.H, "astype"):
             # returns a copy so that downstream can't overwrite the
             # internal attribute. But one can't copy non-arrays
-            xp = array_namespace(self.H)
-            return xp.astype(self.H, self.H.dtype)
+            return self.H.astype(self.H.dtype)
         return self.H
 
 

